emva1288/unittests/test_results.py

```diff
@@ -300,3 +302,66 @@ def test_nans(self):
         self.assertIs(r.DSNU1288_DN(), np.nan)
 
         del r
+
+
+class TestRoutines(unittest.TestCase):
+
+    def setUp(self):
+        # instantiate a bayer filter and a wrong filter
+        bayer_filter = get_bayer_filter(0, 1, 1, 0, 1000, 1000)
+        wrong_filter = np.tile([[1, 1, 1, 0], [1, 0, 0, 1]], (500, 250))
+
+        # Source images. imgc is for color image (image with a bayer filter)
+        self.img = np.abs(np.random.random([1000, 1000]))
+        self.imgc = np.ma.array(self.img, mask=bayer_filter)
+
+        # Image wrong's purpose is to assert that a ValueError is raised
+        self.img_wrong = np.ma.array(self.img, mask=wrong_filter)
+
+        # Pass the highpass filter on the source images
+        res = high_pass_filter(self.img, 5)
+        self.img_filt = res['img']/res['multiplicator']
+        res = high_pass_filter(self.imgc, 5)
+        self.imgc_filt = res['img']/res['multiplicator']
+
+        # To assert that the highpass filter does its job, a low frequency
+        # signal will be introduced in the image
+        signal = np.sin(np.linspace(0, np.pi, 1000))
+        sig_img = [line + val for line, val in zip(self.img, signal)]
+        sig_imgc = np.ma.array(sig_img, mask=bayer_filter)
+
+        # filtering the images with a low frequency signal
+        res = high_pass_filter(sig_img, 5)
+        self.sig_img_filt = res['img']/res['multiplicator']
+        res = high_pass_filter(sig_imgc, 5)
+        self.sig_imgc_filt = res['img']/res['multiplicator']
+
+    def test_highpass_filter(self):
+        # Assert raises in case of bad kernel size or bad mask
+        with self.assertRaises(ValueError):
+            high_pass_filter(self.img_wrong, 5)
+        with self.assertRaises(ValueError):
+            high_pass_filter(self.imgc, 6)
+
+        # assert that the means of the filtered images are almost 0
+        self.assertAlmostEqual(0, np.mean(self.imgc_filt), 3)
+        self.assertAlmostEqual(0, np.mean(self.img_filt), 3)
+
+        # Assert that the std stays about the same after filtering an image
+        # without a low frequency signal
+        self.assertAlmostEqual(np.std(self.img),
+                               np.std(self.img_filt), 1)
+        self.assertAlmostEqual(np.std(self.imgc),
+                               np.std(self.imgc_filt), 1)
+
+        # assert that the means of the filtered images are almost 0 even with
+        # a low frequency signal
+        self.assertAlmostEqual(0, np.mean(self.sig_img_filt), 3)
+        self.assertAlmostEqual(0, np.mean(self.sig_imgc_filt), 3)
+
+        # assert that the std of a the image filtered is almost the same as the
+        # std of that same image with a signal after the filter
+        self.assertAlmostEqual(np.std(self.sig_img_filt),
```

Comment:
The idea is that after filtering, the two images are almost the same, why not assert that the difference of the two filtered images is almost 0?

Reply:
It is in fact more accurate to compare the images element wise. Used np.allclose() to do it, with relative tolerance=0 and absolute tolerance=10e-4. Numpy uses absolute(a - b) <= (atol + rtol * absolute(b)) to compare the elements, tis is why I set rtol to 0. Since the image is generated with np.random, some elements could be very close to 0, so the impact of rtol is not meaningful.